lndr-backend/src/Lndr/Handler/Friend.hs

```diff
@@ -79,8 +81,23 @@ userHandler Nothing Nothing = throwError (err400 {errBody = "No identifying info
 addFriendsHandler :: Address -> [Address] -> LndrHandler NoContent
 addFriendsHandler address friendAddresses = do
     -- TODO verify signature
-    pool <- asks dbConnectionPool
+    (ServerState pool configTVar loggerSet) <- ask
+    config <- liftIO $ readTVarIO configTVar
     liftIO . withResource pool $ Db.addFriends ((address,) <$> friendAddresses)
+    
+    let attemptToNotify friendAddress = do
+            pendingRequests <- liftIO . withResource pool $ Db.lookupPendingRequest address friendAddress
+            when (pendingRequests > 0) $ do
+                pushDataM <- liftIO . withResource pool $ Db.lookupPushDatumByAddress friendAddress
+                nicknameM <- liftIO . withResource pool $ Db.lookupNick address
+
+                forM_ pushDataM $ \(channelID, platform) -> liftIO $ do
+                    responseCode <- sendNotification config (Notification channelID platform nicknameM NewFriendRequest)
+                    let logMsg = "Notification response (" ++ (show friendAddress) ++ "): " ++ show responseCode
+                    liftIO $ pushLogStrLn loggerSet . toLogStr $ logMsg
+    
+    forM_ friendAddresses attemptToNotify
+
```

Comment:
The way I interpret this:
 * After creating a friendship from the `address` to each address in `friendAddresses`, send a notification to each address in `friendAddresses` for which there is no friendship in the reverse direction.

Do we also want to notify when friendships have been mutually established? (i.e: "You and **alice** are now friends.")

If my interpretation is correct, then we can achieve this with two SQL queries instead of N queries by getting the push notification data all at once. Something like…

```sql
SELECT DISTINCT
  n.nickname,
  initiator.friend AS address,
  push.channel_id,
  push.platform
FROM
  friendships initiator
LEFT JOIN
  friendships responder
    ON initiator.friend = responder.origin AND initiator.origin = responder.friend
INNER JOIN
  nicknames n
    ON initiator.origin = n.address
INNER JOIN
  push_data push
    ON initiator.friend = push.address
WHERE
  initiator.origin = ?
    AND
  initiator.friend IN (?)
    AND
  responder.friend IS NULL
```

…where the first given parameter is `address` and the second is the whole `friendAddresses` array. (alternatively it may be be necessary to build the query string manually like `IN (?,?,?,?)`… not sure of the mechanics of prepared statements in this library)

Aside from the performance benefit of consolidating the logic for getting the data needed here, it would also result in the logic here becoming much more straight-forward (i.e: enumerate over the results and send a friend request notification for each one).

Reply:
For fear of over-engineering here, I should add that the changes indicated in this comment are non-blocking, and merely indicate a possible alternative path.